sw/host/cryptotest/cryptotest_parser.rs

```diff
@@ -0,0 +1,904 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Parser for cryptolib test fixtures.
+//!
+//! This file consumes `.c` files and performs a very basic parse on them
+//! to extract the test vector type and any annotations on it.
+//!
+//! This library expects the `.c` file to contain a struct definition
+//! conforming to the following reduced grammar, modulo whitespace.
+//!
+//! ```text
+//! // cryptotest:struct
+//! // Comments...
+//! typedef struct $name? {
+//!   // Comments...
+//!   $type $field;
+//!   // More fields...
+//! } $name;
+//! ```
+//!
+//! Here, `// Comments...` is any number of single-line comments, `$field` is any
+//! valid C identifier, and `$type` is one of the following:
+//! - `bool`
+//! - `uint8_t`, `uint16_t`, `uint32_t`, `uint64_t`
+//! - `int8_t`, `int16_t`, `int32_t`, `int64_t`
+//! - `size_t`
+//! - `const char *` (always a NUL-terminated string).
+//!
+//! Additionally, the following compound types are recognized:
+//! - `Int field[N];`, where `Int` is an integer type above and `N` is a literal.
+//! - `const Int *field;`, for a variable-length array of integers.
+//! - `const char *const *field;`, for a variable-length array of strings.
+//! - `const Int (*field)[N];`, for a variable-length array of fixed arrays of ints.
+//!
+//! Comments that do not begin with `// cryptotest:` are ignored; the rest are parsed
+//! as annotations.
+
+use regex::Regex;
+
+/// A test vector struct.
+///
+/// See the [module docs](self) for information on the grammar.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Struct {
+    /// The name of the struct (as specified by the `typedef`).
+    pub name: String,
+    /// The struct's fields.
+    pub fields: Vec<Field>,
+    /// Any `// cryptotest:` annotations at the top of the struct.
+    pub annots: Vec<Annotation>,
+}
+
+/// A field of a [`Struct`].
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Field {
+    /// The name of the field.
+    pub name: String,
+    /// The type of the field.
+    pub ty: FieldType,
+    /// Any `// cryptotest:` annotations on this struct.
+    pub annots: Vec<Annotation>,
+}
+
+/// A [`Field`]'s type.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum FieldType {
+    /// An array field, represented by a raw `const` pointer. The length of
+    /// the array must be specified by another field of the struct
+    /// via `// cryptotest:len`.
+    Array(Scalar),
+    /// A scalar field, consisting of a single value.
+    Scalar(Scalar),
+}
+
+/// An integer type recognized by cryptotest.
+///
+/// Not only does this include the `stdint.h` types, but it also includes cryptolib
+/// enum types known a priori to the parser.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum Int {
```

Comment:
Ah, does this mean the `Int` from the comment above includes booleans/unsigned ints also?

Reply:
Indeed. I've clarified the text above.